### PR TITLE
[KEYCLOAK-11753] Introduce the ability to add OTP label, when creating OTP credential via 'Configure-OTP' required action or via 'Reset-OTP' authenticator in 'Reset Credentials' flow

### PR DESCRIPTION
**Pull request by iankko**
[KEYCLOAK-11753] Introduce the possibility to add OTP label, when adding OTP code:
    * Through 'Configure-OTP' required action, or
    * Through 'Reset-OTP' authenticator in the 'Reset Credentials' flow

    Add testcases for both of the scenarios

    Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

### CODE REVIEW

**Comment by iankko** [comment]
@mposolda @AlistairDoswald PTAL (once got a chance)

Thanks, Jan

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetCredentialsAlternativeFlowsTest.java
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetCredentialsAlternativeFlowsTest.java]: It seems that "configureResetCredentialsFlowWithoutSendResetEmail" and "configureResetCredentialsFlowWithoutResetPassword" is almost same method? Is it possible to create single method instead of those 2 methods and use the label of the authenticator (EG. "Send Reset Email" or "Reset password") as parameters? Bonus point could be to add more generic method to FlowUtil for remove method by providerId (providerId is better than displayName IMO).

**Review by mposolda — CHANGES_REQUESTED** [review]
I've added one inline comment.

Besides that, I personally don't know what "mnemonic identifier" is and my guess is, that end users might be confused as well. My vote is to:
- keep the label in admin console as "User Label" as it was before
- Keep the term "Device name" in old account console and use this term in the login-config-totp.ftl page as well. It is fine that we keep term "User Label" and then use different term for this for particular authenticators (EG. For TOTP, the "Device name" is maybe the best term I can think about, for WebAuthn, it could be "Authenticator Name" )?

@AlistairDoswald WDYT?

**Comment by iankko** [comment]
> I've added one inline comment.
> 
> Besides that, I personally don't know what "mnemonic identifier" is and my guess is, that end users might be confused as well.

Identifier assisting memory for easier remembering (AKA self-descriptive identifier). But agree, this might not be obvious immediately after reading the label.

> My vote is to:
> 
>     * keep the label in admin console as "User Label" as it was before
> 
>     * Keep the term "Device name" in old account console and use this term in the login-config-totp.ftl page as well. It is fine that we keep term "User Label" and then use different term for this for particular authenticators (EG. For TOTP, the "Device name" is maybe the best term I can think about, for WebAuthn, it could be "Authenticator Name" )?
> 
> 
> @AlistairDoswald WDYT?

Ok, will change it back / revert those changes.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetCredentialsAlternativeFlowsTest.java
Comment (iankko) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetCredentialsAlternativeFlowsTest.java]: Right, both of them doing the same, just disabling different authenticator. Ok, will look into FlowUtil method implementation yet (agree, it would be good to have some method universal enough to be re-usable in different places).

Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetCredentialsAlternativeFlowsTest.java]: @iankko Thanks.

**Comment by mposolda** [comment]
@iankko Thanks

**Comment by iankko** [comment]
> I've added one inline comment.
> 
> Besides that, I personally don't know what "mnemonic identifier" is and my guess is, that end users might be confused as well. My vote is to:
> 
>     * keep the label in admin console as "User Label" as it was before
> 
>     * Keep the term "Device name" in old account console and use this term in the login-config-totp.ftl page as well. It is fine that we keep term "User Label" and then use different term for this for particular authenticators (EG. For TOTP, the "Device name" is maybe the best term I can think about, for WebAuthn, it could be "Authenticator Name" )?
> 
> 
> @AlistairDoswald WDYT?

@mposolda @AlistairDoswald Fixed the conflicts & reverted the previously done `mnemonicIdentifier` change in the PR update.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetCredentialsAlternativeFlowsTest.java
Comment (iankko) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetCredentialsAlternativeFlowsTest.java]: FWIW regarding this one, tried to refactor the routine to be based in `FlowUtil`, but doesn't seem to be able to get it working right now. When trying it (using either methods of `RealmModel`, `RealmResource`, or `AuthenticationManagementResource` getting either `NotSerializableException` or even something like:

`
Caused by: javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.keycloak.models.jpa.entities.AuthenticationExecutionEntity#0b77ab43-105f-443f-ab4f-10b96bf38c06]
`
IOW this will need more time to get refactored into the form of a universal `FlowUtil` method. But not to block this PR, for now, just removed the duplication in the methods and kept it under `ResetCredentialsAlternativeFlowsTest.java` location.

This version WFM now:

```
1:22:10,210 INFO  [org.keycloak.testsuite.arquillian.AuthServerTestEnricher] removing test realms after test class
21:22:10,579 INFO  [org.keycloak.testsuite.arquillian.AuthServerTestEnricher] removed realms: test, 
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 46.641 s - in org.keycloak.testsuite.forms.ResetCredentialsAlternativeFlowsTest
21:22:10,648 INFO  [org.keycloak.testsuite.arquillian.undertow.KeycloakOnUndertow] Stopping auth server.
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
```

and can rewrite it into the `FlowUtil` form later (once got it working).

**Comment by iankko** [comment]
@mposolda @AlistairDoswald This should be ready for re-review again. Besides the `FlowUtil` refactoring, all is done AFAICT.

**Review by mposolda — APPROVED** [review]
@iankko Thanks!

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/RequiredActionTotpSetupTest.java
Comment (AlistairDoswald) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/RequiredActionTotpSetupTest.java:270]: This user is the same as in `setupTotpRegister`. Since it is not cleaned up, it is causing `setupTotpRegister` to fail.
I suggest either creating a different user, or removing it after the test.

**Review by AlistairDoswald — CHANGES_REQUESTED** [review]
There's a small side effect of the setupTotpRegisterVerifyCustomOtpLabelSetProperly test you've added. Once that's corrected, I can merge the PR.

**Comment by AlistairDoswald** [comment]
Closing this PR, following @mposolda suggestion, it will be directly applied to the new "multi-token-prototype-rebased2" branch instead

**Pull request closed without merging**
